Detectors/FIT/common/calibration/include/FITCalibration/FITCalibrationDevice.h

```diff
@@ -0,0 +1,142 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef O2_FITCALIBRATIONDEVICE_H
+#define O2_FITCALIBRATIONDEVICE_H
+
+#include <utility>
+
+#include "FITCalibration/FITCalibrator.h"
+#include "Framework/Task.h"
+#include "Framework/WorkflowSpec.h"
+#include "Framework/DataProcessorSpec.h"
+
+namespace o2::calibration::fit
+{
+
+template <typename InputCalibrationInfoType, typename TimeSlotStorageType, typename CalibrationObjectType>
+class FITCalibrationDevice : public o2::framework::Task
+{
+
+  static constexpr std::size_t OBJECT_SENDING_FREQUENCY = 5;
+  using CalibratorType = FITCalibrator<InputCalibrationInfoType, TimeSlotStorageType, CalibrationObjectType>;
+
+ public:
+  explicit FITCalibrationDevice(std::string inputDataLabel, std::string calibrationObjectPath, int64_t initialTimestamp)
+    : mInputDataLabel(std::move(inputDataLabel)), mCalibrationObjectPath(std::move(calibrationObjectPath)),
+      mInitialTimestamp(initialTimestamp){}
+
+  void init(o2::framework::InitContext& context) final;
+  void run(o2::framework::ProcessingContext& context) final;
+  void endOfStream(o2::framework::EndOfStreamContext& context) final;
+
+ private:
+  void _sendOutputs(o2::framework::DataAllocator& outputs);
+  void _sendOutputsIfStoredEnough(o2::framework::DataAllocator& outputs);
+
+ private:
+  std::string mInputDataLabel;
+  std::string mCalibrationObjectPath;
+  std::unique_ptr<CalibratorType> mCalibrator;
+  int64_t mInitialTimestamp;
+
+};
+
+
+#define FIT_CALIBRATION_DEVICE_TEMPLATES \
+  template <typename InputCalibrationInfoType, typename TimeSlotStorageType, typename CalibrationObjectType>
+
+#define FIT_CALIBRATION_DEVICE_TYPE \
+  FITCalibrationDevice<InputCalibrationInfoType, TimeSlotStorageType, CalibrationObjectType>
+
+
+FIT_CALIBRATION_DEVICE_TEMPLATES
+void FIT_CALIBRATION_DEVICE_TYPE::init(o2::framework::InitContext& context)
+{
+
+  mCalibrator = std::make_unique<CalibratorType>(mInitialTimestamp, mCalibrationObjectPath.c_str());
+  //maybe add later some TF frame options?
+}
+
+FIT_CALIBRATION_DEVICE_TEMPLATES
+void FIT_CALIBRATION_DEVICE_TYPE::run(o2::framework::ProcessingContext& context)
+{
+  auto TFCounter = o2::header::get<o2::framework::DataProcessingHeader*>(context.inputs().get(mInputDataLabel).header)->startTime;
+  auto data = context.inputs().get<gsl::span<InputCalibrationInfoType>>(mInputDataLabel);
+  mCalibrator->process(TFCounter, data);
+  _sendOutputsIfStoredEnough(context.outputs());
```

Comment:
Why? The TimeSlots-based calib. is supposed to compare the distance between the new TF and the edge of every slot currently cached and flush out those slots for which the distance exceeds user-set threshold.

Reply:
I'm not sure if I understand this question.
`_sendOutputsIfStoredEnough` is not related to TF and TimeSlots. I'm just checking here if there is enough produced calibration objects to send them to ccdb-populator.

Reply:
I might have misunderstood your code, but the FITCalibrationDevice runs TimeSlotCalibration-based FITCalibrator, which sends the object to the CCDB populator as soon as the bin is processed (i.e. it has enough stat. and its distance to newly arrived TF exceeds certain threshold).

Reply:
You mean method `finalizeSlot`?
In this implementation, calibration object is not sent to ccdb (ccdb-populator) right after execution of `finalizeSlot`, copy of the calibration object is stored in a vector. Then objects stored in the vector are sent in batches of 5 (can be changed) and also when endOfStream signal arrives.
It can be also changed to sending object right after `finalizeSlot`. I thought it will be more optimial to send those objects in batches.

Reply:
In principle, for the online calibrations we want to have the fresh object in the CCDB asap, so that it can be picked up for processing of subsequent TFs. If this is not needed for your calibrations, then also such a batch mode will work, but is there a reason to introduce an additional layer of caching?

Reply:
Not really, I just added this to improve a bit efficency of the device. I think it will be discussed with FIT experts.

Reply:
Actually we will need the fresh object for online calibration

Reply:
Ok, so I will change it to send calib objects right after calibration.

Reply:
I propose to discuss this at the next calbration review (end of March). Your ideas might be good to looking at amount of code to produce an array of offsets I am little bit scared...